stl/inc/algorithm

```diff
@@ -961,6 +941,322 @@ namespace ranges {
+                    _Pass_fn(_Proj1), _Pass_fn(_Proj2));
+            }
+        }
+
+    private:
+        template <class _It1, class _Se1, class _It2, class _Se2, class _Pr, class _Pj1, class _Pj2>
+        _NODISCARD static constexpr bool _Sized_ranges(_It1 _First1, _Se1 _End1, _It2 _First2, _Se2 _End2,
+            iter_difference_t<_It1> _Count, _Pr _Pred, _Pj1 _Proj1, _Pj2 _Proj2) {
+            _STL_INTERNAL_CHECK(_RANGES distance(_First1, _End1) == _Count);
+            _STL_INTERNAL_CHECK(_RANGES distance(_First2, _End2) == _Count);
+
+            for (;; ++_First1, (void) ++_First2, --_Count) { // Trim matching prefixes
+                if (_Count == 0) { // Everything matched
+                    return true;
+                }
+
+                if (!_STD invoke(_Pred, _STD invoke(_Proj1, *_First1), _STD invoke(_Proj2, *_First2))) { // mismatch
+                    if (_Count == 1) { // Single non-matching characters remain; not a permutation
+                        return false;
+                    }
+
+                    break;
+                }
+            }
+            // If we get here, _Count > 1 and initial elements do not match.
+
+            if constexpr (bidirectional_iterator<_It1> && bidirectional_iterator<_It2>) {
+                // Determine final iterator values
+                auto _Last1 = _Find_last_iterator(_First1, _End1, _Count);
+                auto _Last2 = _Find_last_iterator(_First2, _End2, _Count);
+
+                for (;;) { // Trim matching suffixes
+                    --_Last1;
+                    --_Last2;
+                    if (!_STD invoke(_Pred, _STD invoke(_Proj1, *_Last1), _STD invoke(_Proj2, *_Last2))) { // mismatch
+                        ++_Last1;
+                        ++_Last2;
+                        break;
+                    }
+
+                    if (--_Count == 1) {
+                        return false; // initial elements did not match when trimming prefixes
+                    }
+                }
+                // If we get here, _Count > 1, initial elements do not match, and final elements do not match.
+
+                return _Match_counts(_STD move(_First1), _STD move(_Last1), _STD move(_First2), _STD move(_Last2),
+                    _Pred, _Proj1, _Proj2);
+            } else {
+                return _Match_counts(
+                    _STD move(_First1), _STD move(_End1), _STD move(_First2), _STD move(_End2), _Pred, _Proj1, _Proj2);
+            }
+        }
+
+        template <class _It1, class _Se1, class _It2, class _Se2, class _Pr, class _Pj1, class _Pj2>
+        _NODISCARD static constexpr bool _Unsized_ranges(
+            _It1 _First1, _Se1 _End1, _It2 _First2, _Se2 _End2, _Pr _Pred, _Pj1 _Proj1, _Pj2 _Proj2) {
+            for (;; ++_First1, (void) ++_First2) { // Trim matching prefixes
+                if (_First1 == _End1) { // first range is a (possibly proper?) prefix of the second
+                    return _First2 == _End2;

```

Comment:
I find it very mildly confusing that the comment says "possibly proper?" but that's the reverse of the following `_First2 == _End2` test (when that's equal, the ranges are equal, and the prefix is improper).

How about `// first range is equal to, or is a proper prefix of, the second` ?

Reply:
I think I prefer simply `// first range is a prefix of second` here. "possibly proper" is evident enough without needing to be stated. (Ugh - and what was I thinking with the question mark? The comment seems to be unsure if the first range could be a proper prefix.)